### ISSUE TITLE
VideoCapture::read() returns "true" erroneously

### ISSUE THREAD

**Issue by carmangary** [opencv/opencv#8358]
I am using opencv 3.2 with VS2017 and noticed that there is a problem with VideoCapture and webcams.  If I open my webcam from some other application so that it is displaying video and then do something this:

```
m_webCam = cv::VideoCapture(0);
bool ok = false;
ok = m_webCam.read(m_frame);

```
`ok` is true after this.  I was using opencv 3.1 and Visual Studio 2015 and I can confirm it returned false.  Either moving to opencv 3.2 or Visual Studio 2017 broke the functionality.  Because this returns true, m_frame ends up having uninitialized data and causes my program to hang somewhere within the system code.  If I break my application during the hang it is off in never never land.  Regardless, the read() function should definitely be returning false in this situation.

I also checked and `m_webCam.isOpen` also returns true.

**Comment by sakshamsaxena** [comment]
Hi @carmangary ! I think I'm facing exactly the same issue. 
I've been using Open CV 3.1 for a long time on my little Rasp Pi 2 (last used in Oct '16), and I picked it up yesterday again to tinker with it. I went ahead with a fresh start and reinstalled Open CV 3.2. I was running the _exactly_ same code which I used to run before (process live video from USB webcam), and now it simply quits without error after reading first frame. 

I think it's definitely something related to 3.2, as it's the only difference I made over the months. 

Also, #8471 is also similar I feel. @alalek might know better :sweat_smile:

**Comment by alalek** [comment]
> m_frame ends up having uninitialized data

What exactly do you mean?
cv::Mat can be empty or contains some data.

> my program to hang somewhere within the system code

This may be caused by memory corruption. Try to temporary disable unrelated to OpenCV code to detect real problem.

Try something like this (to dump frame contents into the file):
```
m_webCam = cv::VideoCapture(0);
bool ok = false;
m_frame = cv::Mat::zeros(640, 480, CV_8UC3); // fill zeros for debug purpose
ok = m_webCam.read(m_frame);

cv::FileStorage fs("frame.yml", cv::FileStorage::WRITE);
fs << "frame" << m_frame;
fs.release();
```

or to show it:
```
cv::imshow("frame", m_frame);
cv::waitKey();
```

**Comment by carmangary** [comment]
alalek, thank you for your response.

I supposed I could preinitialize m_frame before calling m_webCam.read(m_frame) and maybe that would help with the hang that is experienced.  But, that really isn't important because the only reason m_frame is being used is because the "read" function returns true when it should be returning false.  However, since it always returns true, I have no way of know whether or not I should do anything with the m_frame or not and that is the real issue.  Even if m_frame is still valid after read is called, I don't want to display it.  

I am able to compile opencv so if I can find time and can figure out how, I will attempt to debug this.  But, it could be awhile  before I can get to that.  I'll try initializing the frame as you suggested in my code but as I mentioned, that doesn't really help in the grand scheme of things.  The root problem is that the read method returns the wrong status message.

**Comment by carmangary** [comment]
I did a quick test where I ran something like this:

```
m_webCam = cv::VideoCapture(0);
m_frame = cv::Mat::zeros(640, 480, CV_8UC3); // fill zeros for debug purpose
ok = m_webCam.read(m_frame);
m_frame = cv::Mat::zeros(640, 480, CV_8UC3); // fill zeros for debug purpose
ok = m_webCam.read(m_frame);
```

The first call to m_webCam.read() returns true even though my camera is in use by another application.
The second call to m_webCam hangs.

**Comment by alalek** [comment]
@carmangary Thank you!
Could you please dump application output with [OPENCV_VIDEOCAPTURE_DEBUG=1](https://github.com/opencv/opencv/blob/3.2.0/modules/videoio/src/cap.cpp#L127) environment variable?
This should show used Video backend in your configuration.

**Comment by carmangary** [comment]
@alalek I added the enviroment variable but I don't know where to find the debug output.  I looked in the application directory.  Where is it stored?

**Comment by alalek** [comment]
It should be dumped on stderr via `fprintf(stderr, ...)`.

**Comment by carmangary** [comment]
@alalek  I found this in my debug output.  Is this what you want?  If not, let me know and I will try again later when I have more time.

***** VIDEOINPUT LIBRARY - 0.1995 - TFW07 *****

SETUP: Microsoft® LifeCam HD-5001
SETUP: Couldn't find preview pin using SmartTee
SETUP: Default Format is set to 640x480
SETUP: trying specified format RGB24 @ 640x480
SETUP: trying format RGB24 @ 640x480
SETUP: trying format RGB32 @ 640x480
SETUP: trying format RGB555 @ 640x480
SETUP: trying format RGB565 @ 640x480
SETUP: trying format YUY2 @ 640x480
SETUP: Capture callback set

ERROR: Could not start graph

SETUP: Disconnecting device 0
SETUP: freeing Grabber Callback
ERROR - Could not pause pControl
SETUP: freeing Renderer
SETUP: freeing Capture Source
SETUP: freeing Grabber Filter
SETUP: freeing Grabber
SETUP: freeing Control
SETUP: freeing Media Event
SETUP: freeing Media Type
SETUP: removing filter Microsoft® LifeCam HD-5001...
SETUP: filter removed Microsoft® LifeCam HD-5001

SETUP: freeing Capture Graph
SETUP: freeing Main Graph
SETUP: Device 0 disconnected and freed

**Comment by alalek** [comment]
Not exactly. But it helps - this output is from [dshow](https://github.com/opencv/opencv/blob/3.2.0/modules/videoio/src/cap_dshow.cpp#L2586) backend.

**Comment by carmangary** [comment]
@alalek Do you need anything else from me?
